File: c/src/p0004.c
/*
Project Euler Problem 4

This was pretty easy to do, given the digit infrastructure I already had built up

Problem:

A palindromic number reads the same both ways. The largest palindrome made from
the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
*/
#ifndef EULER_P0004
#define EULER_P0004
#include <stdint.h>
#include <inttypes.h>
#include <stdio.h>
#include "include/macros.h"
#include "include/digits.h"

uint32_t EMSCRIPTEN_KEEPALIVE p0004() {
    uint32_t answer = 0, i, j, a, z, prod;
    bool broken;
    digit_counter dc;
    for (i = 100; i < 1000; i++)
        for (j = 100; j < 1000; j++) {
            prod = i * j;
            dc = digits(prod);
            broken = false;
            for (a = 0, z = dc.idx; a < z; a++, z--)
                if (dc.digits[a] != dc.digits[z]) {
                    broken = true;
                    break;
                }
            if (!broken)
                answer = max(answer, prod);
            free_digit_counter(dc);
        }
    return answer;
}

PROGRAM_TAIL(PRIu32, p0004)
#endif

File: fortran/src/p0004.f90
! Project Euler Problem 4
!
! I modeled this after the C++ solution, because it was by far the simplest
!
! Problem:
!
! A palindromic number reads the same both ways. The largest palindrome made from
! the product of two 2-digit numbers is 9009 = 91 × 99.
!
! Find the largest palindrome made from the product of two 3-digit numbers.

module Problem0004
    implicit none
contains
    pure integer function p0004() result(answer)
        integer :: i, j, k, n, prod, length
        logical :: is_palindrome
        character(len=8) :: string

        answer = 0
        do i = 100, 999
            do j = 100, 999
                prod = i * j
                is_palindrome = .true.
                write(string, '(I0)') prod
                length = len_trim(string)

                do k = 1, length
                    n = length - k + 1
                    if (string(k:k) /= string(n:n)) then
                        is_palindrome = .false.
                        exit
                    end if
                end do

                if (is_palindrome) then
                    answer = max(answer, prod)
                end if
            end do
        end do
    end function p0004
end module Problem0004
